zyte_api/_retry.py

```diff
@@ -177,3 +177,62 @@ def build(self) -> AsyncRetrying:
 
 
 zyte_api_retrying: AsyncRetrying = RetryFactory().build()
+
+
+def is_maybe_temporary_error(exc: BaseException) -> bool:
+    return (
+        isinstance(exc, RequestError)
+        and exc.status >= 500
+        and exc.status not in {503, 520}
+    )
+
+
+class ConservativeRetryFactory(RetryFactory):
+    """Alternative factory class that builds :data:`conservative_retrying`.
+
+    To create a custom retry policy based on :data:`conservative_retrying`, you
+    can subclass this factory class, modify it as needed, and then call
+    :meth:`build` on your subclass to get the corresponding
+    :class:`tenacity.AsyncRetrying` object.
+
+    For example, to increase the maximum number of attempts for errors treated
+    as temporary download errors by :data:`conservative_retrying` from 16 (i.e.
+    15 retries) to 32 (i.e. 31 retries):
+
+    .. code-block:: python
+
+        from tenacity import stop_after_attempt
+        from zyte_api import ConservativeRetryFactory
+
+
+        class CustomRetryFactory(ConservativeRetryFactory):
+            temporary_download_error_stop = stop_after_attempt(32)
+
+
+        CUSTOM_RETRY_POLICY = CustomRetryFactory().build()
+    """
+
+    retry_condition = RetryFactory.retry_condition | retry_if_exception(
+        is_maybe_temporary_error
+    )
+
+    temporary_download_error_stop = stop_after_attempt(16)
+
+    def stop(self, retry_state: RetryCallState) -> bool:
+        assert retry_state.outcome, "Unexpected empty outcome"
+        exc = retry_state.outcome.exception()
+        assert exc, "Unexpected empty exception"
+        if is_maybe_temporary_error(exc):
+            return self.temporary_download_error_stop(retry_state)
```

Comment:
I must admit I don't know enough about these 500x+ errors. 
Should they use the same stopping & delay logic as 520s? Should they share the state as well? E.g. you get 520, then 520, then 500 - should 500 be considered a 3rd retry, or 1st? Should we wait more for 500s, or less, or the same as for 520s? Should we group 520 and 521s in this policy, but have other 500s handled separately?

Reply:
> Should they share the state as well? E.g. you get 520, then 520, then 500 - should 500 be considered a 3rd retry, or 1st?

This is a great question. I am quite unfamiliar with tenacity, and assumed the state is always shared, hence I went with “all on the same sack” to avoid a scenario where 520 is retried 16 times, 500 3 times, and a 520→520→520→500 sequence exceeds retries. I’ll have a look at tenacity docs about this.

Reply:
While [tenacity’s state object does not provide means to track different error sets by default](https://github.com/jd/tenacity/blob/cb15300d9b4358d9bdb6cfd33d433368700b3abe/tenacity/__init__.py#L513-L545), this is Python, so we could extend that object to track different retry groups.

What about this?
- 429 and 503 keep working as usual.
- 520 is retried as usual, with the proposed extra retries (32 attempts), but we track it, so 429→429→520→429→520 only counts as 2 attempts / 1 retry.
- Similar with network errors, where we ensure that the “stop after 15 minutes” means “stop after 15 minutes of network errors in a row”, so that 15 minutes of 429 errors followed by a network error do not exhaust retries, and network error followed by 30 minutes of 429 followed by another network error does not exhaust retries either.
- I’m hesitant about putting 521 errors in the same bag as 520, as I feel they should have fewer retries, both in theory (they are meant to indicate a permanent download error) and in practice (they represent 4.5% of 520 errors for the past 90 days). Although the stats may be misleading, it could be that on some websites you are getting 521 instead of 520 always, but fewer websites are in that scenario. In any case, I would suggest fewer attempts (8?) for those, and individual tracking. Maybe we could make them count as 520 but not the other way around, e.g. 521*8 exhausts retries (regardless of in-between 521s), 520*16→521*7→520*9 exhausts 520’s 32 attempts.
- For all other 5xx, they should seldom happen, so we could go for 4 attempts, tracking each code individually, and failing only if the attempts happen in a row, excluding rate limiting responses. e.g. 500*3 →  429*10 → 500 fails. 500 → 502 counts as 1 attempt (for 502), 500 → 502 → 500 counts as 1 attempt (for 500).

How does that sound?

I wonder if individual tracking on its own could make it unnecessary to increase retries so much.

I also wonder if we should try running some Scrapy jobs that needed a more aggressive retry strategy with the one we prepare here, and see how they fare.

And, as we add extra tracking, I wonder if we should try and modify the error message that is logged when retries are exceeded, to log the chain of status codes that caused the retry exhaustion. Maybe even with the datetime when each status code was received.

Reply:
> Hm, the jump from 16 to 32 is not that big. I haven't realized we have 16 by default.

[We have 4](https://github.com/zytedata/python-zyte-api/blob/9ed623d922f5027bce06d2b7df58a96d5c64a717/zyte_api/_retry.py#L140), not 16, so it is indeed a big jump (8×). It is based on actual custom retry policies I saw (many using 32, a few up to 40 attempts). However, this is where this though comes in:

> I wonder if individual tracking on its own could make it unnecessary to increase retries so much.

So, we could start with a lower jump (4 → 8?), get people to try it out, and see if thanks to the new error tracking logic such a jump is enough.